src/coreclr/jit/morph.cpp

```diff
@@ -11328,15 +11318,104 @@ GenTree* Compiler::fgOptimizeBitwiseAnd(GenTreeOp* andOp)
     GenTree* op1 = andOp->gtGetOp1();
     GenTree* op2 = andOp->gtGetOp2();
 
-    // Fold "cmp & 1" to just "cmp".
-    if (andOp->TypeIs(TYP_INT) && op1->OperIsCompare() && op2->IsIntegralConst(1))
+    // X & X -> X. Intentionally enabled only for cheap trees for better throughput.
+    if ((op2->IsLocal() || op2->IsInvariant()) && GenTree::Compare(op1->gtEffectiveVal(), op2))
     {
-        DEBUG_DESTROY_NODE(op2);
-        DEBUG_DESTROY_NODE(andOp);
+        return op1;
+    }
```

Comment:
This looks like it doesn't handle `GTF_REVERSE_OPS` correctly (the side effects may modify `op2` if it's an address exposed local in that case).

Nit: The comment is a bit misleading -- it wouldn't be legal to do this for other trees without more complex checks (e.g. for obvious reasons you couldn't fold `M() & M() => M()` even though `GenTree::Compare` would return true).